### DISCUSSION TITLE
fit the map area to show the road between two points

### DISCUSSION THREAD

**Discussion by jesussmile** [liodali/osm_flutter#137] [Show and tell]
I was just working with your new method `limitAreaMap` and with it I am able to fit the map with the selection of roads,
First I took the coordinates of the two roads and then found a midpoint. Then using `BoundingBox` I am able to resize the map to fit the road
```ruby
getMidPointLatLong(GeoPoint latLng1, GeoPoint latLng2) async {
    double midLat, midLong;

    int coefficient = 1;

    if ((latLng1.latitude > 90 && latLng2.latitude < -90) ||
        (latLng2.latitude > 90 && latLng1.latitude < -90)) {
      midLat = 180 - ((latLng1.latitude - latLng2.latitude) / 2).abs();

      if ((latLng1.latitude < 0 &&
              latLng1.latitude.abs() < latLng2.latitude.abs()) ||
          (latLng2.latitude < 0 &&
              latLng2.latitude.abs() < latLng1.latitude.abs()))
        coefficient = -1;

      midLat *= coefficient;
    } else
      midLat = (latLng2.latitude + latLng1.latitude) / 2;

    coefficient = 1;

    if ((latLng1.longitude > 90 && latLng2.longitude < -90) ||
        (latLng2.longitude > 90 && latLng1.longitude < -90)) {
      midLong =
          180 - ((latLng1.longitude.abs() - latLng2.longitude.abs()) / 2).abs();

      if ((latLng1.longitude < 0 &&
              latLng1.longitude.abs() < latLng2.longitude.abs()) ||
          (latLng2.longitude < 0 &&
              latLng2.longitude.abs() < latLng1.longitude.abs()))
        coefficient = -1;

      midLong *= coefficient;
    } else
      midLong = (latLng2.longitude + latLng1.longitude) / 2;

    print("middle ${GeoPoint(latitude: midLat, longitude: midLong)}");
    
    await osmController.limitAreaMap(BoundingBox(
        north: midLat, east: midLong, south: midLat, west: midLong));
    await osmController.zoom(-4);
    await osmController.removeLimitAreaMap();
  }
```
Not sure if its anyuseful to this project but still thought I would share, 
if you want to be more precise than zooming out and control then use this
it based on distance i.e Kilometers 
```ruby
kmInDegree(lat, long) {
    const pi = Math.pi;
    const eSq = 0.00669437999014;
    const a = 6378137.0; //equatorial radius in metres
    lat = lat * pi / 180; //convert to radians
    long = long * pi / 180; //convert radians

    var latLength = (pi * a * (1 - eSq)) /
        (180 * Math.pow((1 - eSq * Math.pow(Math.sin(lat), 2)), 3 / 2));

    var longLength = (pi * a * Math.cos(long)) /
        (180 * Math.sqrt((1 - (eSq * Math.pow(Math.sin(long), 2)))));

    return (latLength, longLength);
  }
```
https://user-images.githubusercontent.com/11044978/127452743-dc942ad3-ee27-4d4f-8aca-d159d779ae07.mp4

**Comment by liodali** [comment]
thanks for sharing, it will be usefull we can add it as completary function
i will add it soon
for `kmInDegree` can explain utility of it ?

**Reply by jesussmile** [reply]
`kmInDegree` will convert kilometers to degrees, you will use it before sending it to `BoundingBox`, after getting the midpoint. for example, if you return this 
```ruby
return {
    lat: 1000 / latLength,
    long: 1000 / longLength
  };
```
It will add 1 km in degrees to east, west north, and south.  For smaller distances, the `zoom` method of your plugin will work fine but if the distance is large or very small then you need to add the kminDegree value to it so that the road will fit in the map, or you can fit anything like multiple Geopoint, etc, This will guarantee 100% that all markers, location or road is visible in the screen. I hope it makes sense.

**Reply by liodali** [reply]
also you should know that i will add zoomToBoundingBox ( available in all platform ) that it will get all area visibile

**Reply by jesussmile** [reply]
Great! That would be much better. Right now everything works fine for me but  I am stuck on two things. Rotate the icon and replace the default icon.  Any update on this ?

**Reply by liodali** [reply]
for now I'm little bit focing on making ios support stable and i started adding the web support (still prototyping)  but  i will add rotate static point this weekend and i will add default icon next weekend

**Comment by renehw** [comment]
This recurse was avaliable in the last version?

**Comment by liodali** [comment]
i will add fit area camera method in general soon i will add it in 0.28+

**Comment by renehw** [comment]
Fine, thanks for this. For now, I implemented the solution that Jesus send for the get the midle point, but I can't getter the currect zoom, you may help me with this?

**Reply by liodali** [reply]
for the zoom it's tricky you can for example base on distance to calculate it
for example for each level in the  zoom there are specific number distnce  for each tile 
* example for zoom level 18 we have distance of 0.5m/pixel (https://wiki.openstreetmap.org/wiki/Zoom_levels)
so you can make approximatly if distance > X km zoom = 8 etc ..

this implementation of zoomToboundingbox in osmdroid 
- https://github.com/osmdroid/osmdroid/blob/4094f3869a72215e5f7fb5ead78c21674b60ab07/osmdroid-android/src/main/java/org/osmdroid/views/MapView.java#L536
- https://github.com/osmdroid/osmdroid/blob/4094f3869a72215e5f7fb5ead78c21674b60ab07/osmdroid-android/src/main/java/org/osmdroid/util/TileSystem.java#L410

**Reply by renehw** [reply]
I Understand, I will try implement this, thank you!
